sshd-core/src/main/java/org/apache/sshd/common/kex/MontgomeryCurve.java

```diff
@@ -0,0 +1,158 @@
+        this.algorithm = algorithm;
+        this.keySize = keySize;
+        this.digestFactory = digestFactory;
+        this.encodedPublicKeyPrefix = encodedPublicKeyPrefix;
+        encodedKeySize = keySize + encodedPublicKeyPrefix.length;
+        boolean supported;
+        KeyPairGenerator generator = null;
+        KeyFactory factory = null;
+        try {
+            SecurityUtils.getKeyAgreement(algorithm);
+            generator = SecurityUtils.getKeyPairGenerator(algorithm);
+            factory = SecurityUtils.getKeyFactory(algorithm);
+            supported = true;
+        } catch (GeneralSecurityException ignored) {
+            supported = false;
+        }
+        this.supported = supported && digestFactory.isSupported();
+        keyPairGenerator = generator;
+        keyFactory = factory;
+    }
+
+    public String getAlgorithm() {
+        return algorithm;
+    }
+
+    @Override
+    public int getKeySize() {
+        return keySize;
+    }
+
+    @Override
+    public boolean isSupported() {
+        return supported;
+    }
+
+    public KeyAgreement createKeyAgreement() throws GeneralSecurityException {
+        return SecurityUtils.getKeyAgreement(algorithm);
+    }
+
+    public Digest createDigest() {
+        return digestFactory.create();
+    }
+
+    public KeyPair generateKeyPair() {
+        return keyPairGenerator.generateKeyPair();
+    }
+
+    public byte[] encode(PublicKey key) throws InvalidKeyException {
+        return extractSubjectPublicKey(key.getEncoded());
+    }
+
+    public PublicKey decode(byte[] key) throws InvalidKeySpecException {
+        if (key.length < getKeySize()) {
+            throw new InvalidKeySpecException("Provided key is too small for " + getAlgorithm());
+        }
+        // ideally, we'd just parse the key as a BigInteger and then create a XECPublicKeySpec in Java 11
+        // BouncyCastle supports a separate API, so we can use the generic X.509 encoding scheme supported by both
+        byte[] encoded = new byte[encodedKeySize];
+        System.arraycopy(encodedPublicKeyPrefix, 0, encoded, 0, encodedPublicKeyPrefix.length);
+        // note that key can be either the raw key data or it may be prefixed by a padding byte and the key length.
+        // these two bytes are already present as the last two bytes in encodedPublicKeyPrefix, thus there is no harm
+        // in potentially overwriting it
+        System.arraycopy(key, 0, encoded, encodedKeySize - key.length, key.length);
+        return keyFactory.generatePublic(new X509EncodedKeySpec(encoded));
+    }
+
+    private static byte[] extractSubjectPublicKey(byte[] subjectPublicKeyInfo) throws InvalidKeyException {
+        try {
+            //  SubjectPublicKeyInfo ::= SEQUENCE {
+            //   algorithm AlgorithmIdentifier,
+            //   subjectPublicKey BIT STRING }
+            ASN1Object spki = new DERParser(subjectPublicKeyInfo).readObject();
```

Comment:
I found some easier ways to encode/decode keys in a separate project I'm working on: https://github.com/o1c-dev/o1c/blob/2e41f8e6268ead61562474424a01f256a7afe8e1/java8/src/main/java/dev/o1c/bc/BouncyCastleKeyPairCodec.java#L75

Reply:
- [ ] Just extract the key data directly from the byte array as it's the last `keySize` bytes in the X.509 encoded form

Reply:
Done.